src/cxx.cc

```diff
@@ -132,6 +136,34 @@ String::String(const char16_t *s, std::size_t len) {
              len);
 }
 
+String String::lossy(const std::string &s) noexcept {
+  return String::lossy(s.data(), s.length());
+}
+
+String String::lossy(const char *s) noexcept {
+  assert(s != nullptr);
+  return String::lossy(s, std::strlen(s));
+}
+
+String String::lossy(const char *s, std::size_t len) noexcept {
+  assert(s != nullptr || len == 0);
+  String string;
```

Comment:
This line calls the rust::String default constructor which constructs a Rust String in this location, but then the line after this treats it as uninitialized and writes a brand new String on top of the first one using ptr::write. The original string is leaked.

Ideally we wouldn't default construct the first one in the first place, since that is a whole unneeded round-trip from C++ to Rust. Could you look into making rust::String::lossy go through a private constructor of rust::String instead? Something like `return String(lossy_t{}, s, len)` where the constructor is `private:` `String(lossy_t, const char *s, std::size_t len) noexcept`. That way it'll work like the existing constructors with no default constructor involved.

Reply:
Oof, great points, thanks. Done as you suggest.